### PR TITLE
[Country] don't call the request based resolvers every time

### PR DESCRIPTION
**Pull request by dpfaffenbauer**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #1260

### CODE REVIEW

File: src/CoreShop/Component/Address/Context/RequestBased/CountryContext.php [hunk 1]
```diff
@@ -46,7 +51,13 @@ public function __construct(RequestResolverInterface $requestResolver, RequestSt
     public function getCountry()
     {
         try {
-            return $this->getCountryForRequest($this->getMasterRequest());
+            $request = $this->getMasterRequest();
+
+            if (null !== $this->country) {
+                return $this->country;
```
Comment (solverat) [inline src/CoreShop/Component/Address/Context/RequestBased/CountryContext.php]: In currency resolver this is dedicated to a cache resolver. Just my two cents. :) 
However, it's working!

**Comment by solverat** [comment]
@dpfaffenbauer i think we also should modify the `GeoLiteBasedRequestResolver` => `findCountry` Method. On `dev`, the `checkIfIpIsPrivate` methods always returns true. Since this is very late (Reader has been opened already) we have a lot of unnessary IO traffic. My proposal:

```php
public function findCountry(Request $request)
{
    $geoDbFile = PIMCORE_CONFIGURATION_DIRECTORY . '/GeoLite2-City.mmdb';
    $record = null;

    $clientIp = $request->getClientIp();
    if ($this->checkIfIpIsPrivate($clientIp)) {
        throw new CountryNotFoundException();
    }

    if (!file_exists($geoDbFile)) {
        throw new CountryNotFoundException();
    }

    try {
        $reader = new Reader($geoDbFile);
        $record = $reader->city($clientIp);
        $country = $this->countryRepository->findByCode($record->country->isoCode);
        if ($country instanceof CountryInterface) {
            return $country;
        }
    } catch (\Exception $e) {
        //If something goes wrong, ignore the exception and throw a CountryNotFoundException
    }

    throw new CountryNotFoundException();
}
```

File: src/CoreShop/Component/Address/Context/RequestBased/CountryContext.php [hunk 2]
```diff
@@ -46,7 +51,13 @@ public function __construct(RequestResolverInterface $requestResolver, RequestSt
     public function getCountry()
     {
         try {
-            return $this->getCountryForRequest($this->getMasterRequest());
+            $request = $this->getMasterRequest();
```
Comment (jdreesen) [inline src/CoreShop/Component/Address/Context/RequestBased/CountryContext.php]: Why do you need this variable? It is only used in `$this->getCountryForRequest($request)` (after the `if`)

Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/CountryContext.php]: Cause it throws an exception if no request is available

**Comment by dpfaffenbauer** [comment]
@solverat can you maybe adjust the PR?

File: src/CoreShop/Component/Address/Context/RequestBased/CountryContext.php (hunk 1 above)
Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/CountryContext.php]: ok, it is now separated

**Comment by dpfaffenbauer** [comment]
@solverat changed to store the geo-lite result in cache

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php [hunk 3]
```diff
@@ -25,43 +26,89 @@ final class GeoLiteBasedRequestResolver implements RequestResolverInterface
...
+            throw new CountryNotFoundException();
+        }
+
+        $country = $this->countryRepository->findByCode($countryIsoCode);
+
+        if (!$country instanceof CountryInterface) {
+            throw new CountryNotFoundException();
+        }
+
+        $this->cache->save($cacheKey, $countryIsoCode);
```
Comment (solverat) [inline src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: To make this official, I don't like the idea of adding this the Cache. However, two improvement proposals from my side:
- Use a short TTL (like 1h)
- Use a cacheprefix to keep it more transparent. Like `sprintf('client_ip2c_%s', md5($ip));`

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php [hunk 4]
```diff
@@ -25,43 +26,89 @@ final class GeoLiteBasedRequestResolver implements RequestResolverInterface
...
         }
 
-        throw new CountryNotFoundException();
+        $countryIsoCode = $this->guessCountryByGeoLite($clientIp);
+
+        if ($countryIsoCode === null) {
+            throw new CountryNotFoundException();
+        }
+
+        $country = $this->countryRepository->findByCode($countryIsoCode);
```
Comment (solverat) [inline src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: Reduce `findByCode` calls to one.

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php [hunk 5]
```diff
@@ -25,43 +26,89 @@ final class GeoLiteBasedRequestResolver implements RequestResolverInterface
...
+        if ($countryIsoCode = $this->cache->getItem($cacheKey)) {
+            $country = $this->countryRepository->findByCode($countryIsoCode);
 
-                    if ($country instanceof CountryInterface) {
-                        return $country;
-                    }
-                }
-            } catch (\Exception $e) {
-                //If something goes wrong, ignore the exception and throw a CountryNotFoundException
+            if ($country) {
```
Comment (solverat) [inline src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: use `instanceof`

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php (hunk 4 above)
Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: How do you mean? introducing another function?

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php (hunk 5 above)
Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: done

File: src/CoreShop/Component/Address/Context/RequestBased/CachedCountryContext.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findCountry(Request $request)
+    {
+        if (null === $this->country) {
+            $this->country = $this->inner->findCountry($request);
+
+            return $this->country;
```
Comment (jdreesen) [inline src/CoreShop/Component/Address/Context/RequestBased/CachedCountryContext.php]: This return is not really necessary.

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php [hunk 7]
```diff
@@ -25,43 +26,89 @@ final class GeoLiteBasedRequestResolver implements RequestResolverInterface
...
-        $record = null;
+        if (!file_exists($this->geoDbFile)) {
+            throw new CountryNotFoundException();
+        }
 
-        if (file_exists($geoDbFile)) {
-            try {
-                $reader = new Reader($geoDbFile);
+        $record = null;
+        $isoCode = null;
```
Comment (jdreesen) [inline src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: This variable is unused

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php (hunk 3 above)
Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: More like a day at least, IP address don't change that often.... even if they do, the new address probably results for the same country, right?

Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: done

File: src/CoreShop/Component/Address/Context/RequestBased/CachedCountryContext.php (hunk 6 above)
Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/CachedCountryContext.php]: done

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php (hunk 7 above)
Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: done

**Comment by dpfaffenbauer** [comment]
@solverat @jdreesen happy now?

File: src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php [hunk 8]
```diff
@@ -25,43 +26,87 @@ final class GeoLiteBasedRequestResolver implements RequestResolverInterface
...
     /**
      * @param CountryRepositoryInterface $countryRepository
+     * @param CoreHandlerInterface       $cache
      */
-    public function __construct(CountryRepositoryInterface $countryRepository)
+    public function __construct(CountryRepositoryInterface $countryRepository, CoreHandlerInterface $cache)
     {
         $this->countryRepository = $countryRepository;
+        $this->cache = $cache;
+        $this->geoDbFile = PIMCORE_CONFIGURATION_DIRECTORY . '/GeoLite2-City.mmdb';
```
Comment (jdreesen) [inline src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: As this property is private, and there's (currently) no way to change it, why not make it a `const` instead? 

```php
private const GEO_DB_FILE = PIMCORE_CONFIGURATION_DIRECTORY . '/GeoLite2-City.mmdb';
```

Comment (dpfaffenbauer) [reply src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php]: no point in that, Pimcore stores the file there as well

**Pull request merged**